### PR TITLE
[Ray] Reconstruct worker

### PR DESCRIPTION
**Pull request by fyrestone**
## What do these changes do?

- Add the `ClusterAPI.reconstruct_worker` API.
- Implement `ClusterAPI.reconstruct_worker` for ray.
- New ray actor pool states: INIT, POOL_READY, SERVICE_READY.
- Move `FaultInjectionSubtaskProcessor` to fault_injection_patch.py
- More reliable kill and wait logic for ray.
- Add `test_rerun_subtask_describe` case.
- Log exceptions in `monitor_sub_pools` instead of stop monitoring silently.
- Add the `register_slot` API for `BandSlotControlActor`.
- Fix one slot may be assigned more than one subtask when rerun subtask.

### CODE REVIEW

File: mars/services/scheduling/worker/execution.py [hunk 1]
```diff
@@ -333,10 +333,8 @@ async def _run_subtask_once():
                     logger.debug('Wait for subtask to cancel timed out (%s). '
                                  'Start killing slot %d', subtask_info.kill_timeout, slot_id)
                     await slot_manager_ref.kill_slot(slot_id)
-                    # as slot is restarted, we does not need to trace it
                     sub_pool_address = await slot_manager_ref.get_slot_address(slot_id)
                     await mo.wait_actor_pool_recovered(sub_pool_address, self.address)
-                    slot_id = None
```
Comment (wjsi) [inline mars/services/scheduling/worker/execution.py:339]: Why is this removed here? Removing this will cause an unexpected `release_free_slot` call and may have unexpected consequences.

File: mars/services/scheduling/worker/workerslot.py [hunk 2]
```diff
@@ -151,6 +154,16 @@ def release_free_slot(self, slot_id: int, pid: Optional[int] = None):
...
 
+        if pid is not None:
+            if slot_id in self._slot_to_session_stid:
+                # We should release the slot by one role, if the slot is
+                # acquired by the SubtaskExecutionActor, then the slot
+                # should be released by it, too.
+                session_stid = self._slot_to_session_stid[slot_id]
+                logger.info('Slot %s released by pid %s, but not for %s',
+                            slot_id, pid, session_stid)
+                return
```
Comment (wjsi) [inline mars/services/scheduling/worker/workerslot.py:165]: This section of code is a little confusing here. I would prefer adding a separate actor API maintaining slot-pid bindings.

File: mars/deploy/oscar/tests/test_fault_injection.py [hunk 3]
```diff
@@ -48,17 +51,25 @@ async def fault_cluster(request):
...
+            return self._fault_count
+
         def get_fault(self, pos: FaultPosition, ctx=None) -> FaultType:
+            # Check op types if fault_op_types provided.
+            if fault_op_types and type(ctx.get('operand')) not in fault_op_types:
+                return FaultType.NoFault
             if self._fault_count.get(pos, 0) > 0:
                 self._fault_count[pos] -= 1
+                if fault_op_types:
+                    print(f"get_fault {fault_type} {pos} {ctx}")
```
Comment (wjsi) [inline mars/deploy/oscar/tests/test_fault_injection.py:72]: Is this line for debug?

**Review by wjsi — COMMENTED** [review]
Besides all these comments, I'm wondering who is the caller of `reconstruct_worker`. I don't think it is for the user end as it is not possible to ask an end user to interfere an individual worker in the cluster.

File: mars/deploy/oscar/ray.py [hunk 4]
```diff
@@ -162,22 +168,72 @@ async def new_worker(self, worker_address, band_to_slot=None):
...
+
+            if state == RayPoolState.INIT:
+                await actor.start.remote()
+            else:
+                assert state == RayPoolState.POOL_READY
+
+            start_time = time.time()
+            await start_worker(address, self.address, self._band_to_slot, config=self._config)
+            await actor.mark_service_ready.remote()
+            logger.info('Start services on worker %s succeeds in %.4f seconds.',
```
[the remark this answers is not in the document]
Comment (fyrestone) [reply mars/deploy/oscar/ray.py:230]: This is same log message and log level as in `new_worker`.

File: mars/deploy/oscar/tests/test_fault_injection.py (hunk 3 above)
Comment (fyrestone) [reply mars/deploy/oscar/tests/test_fault_injection.py:72]: It's for the fault injection test case to know which fault types are raising. But, I am OK to remove it.

File: mars/oscar/backends/pool.py [hunk 5]
```diff
@@ -1029,19 +1027,26 @@ async def monitor_sub_pools(self):
...
+                            if self._auto_recover:
+                                await self.recover_sub_pool(address)
+                                if self._on_process_recover is not None:
+                                    self._on_process_recover(self, address)
+                        if recover_events_discovered:
+                            event = self._recover_events.pop(address)
+                            event.set()
+                    except asyncio.CancelledError:
+                        raise
+                    except Exception:
```
[the remark this answers is not in the document]
Comment (fyrestone) [reply mars/oscar/backends/pool.py:1046]: The original logic is to ignore the exceptions and stop monitoring sub pools, because no one await the task.(the monitor task fails silently). This change is to log the exceptions and continue monitoring sub pools, I think it's better than fails silently.

File: mars/services/scheduling/worker/execution.py (hunk 1 above)
Comment (fyrestone) [reply mars/services/scheduling/worker/execution.py:339]: Because the original logic has two roles to release the free slot, They may release one slot twice when fault occurred.
- The `BandSlotControlActor`
- The `SubtaskExecutionActor`

This PR makes the acquire and release logic in one role, the `SubtaskExecutionActor`. The `BandSlotControlActor` just register the slots, it does not actually release the slot.

The corner case is that, when the `SubtaskExecutionActor` get an error when executing subtask. e.g. the a `MarsError`, but the sub pool process crashes after the `MarsError`. The `SubtaskExecutionActor` don't know if the sub pool process really crashes. So, the slots may be released twice. Another fix is to kill the sub pool when any exceptions catched, but it does not fix the root cause: `two roles releasing slots`.

File: mars/services/scheduling/worker/workerslot.py (hunk 2 above)
Comment (fyrestone) [reply mars/services/scheduling/worker/workerslot.py:165]: I prefer adding a `register_slot` API for the `BandSlotControlActor`.

**Comment by fyrestone** [comment]
> Besides all these comments, I'm wondering who is the caller of `reconstruct_worker`. I don't think it is for the user end as it is not possible to ask an end user to interfere an individual worker in the cluster.

The caller of `reconstruct_worker` should be a monitor logic in the supervisor. it's an API to operate the cluster worker, so I think it can be placed next to the `request_worker` and `release_worker`.

File: mars/deploy/oscar/ray.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fyrestone) [reply mars/deploy/oscar/ray.py:230]: Please refer to https://github.com/mars-project/mars/blob/master/mars/deploy/oscar/ray.py#L165

File: mars/services/scheduling/worker/execution.py (hunk 1 above)
Comment (fyrestone) [reply mars/services/scheduling/worker/execution.py:339]: Please refer to this test case: https://github.com/mars-project/mars/pull/2413/files#diff-b233da94b04ddd71d61754e6c850c2d25a97087d2f5cc62e6da40c13a31bfc16R241

File: mars/deploy/oscar/ray.py [hunk 6]
```diff
@@ -162,22 +168,72 @@ async def new_worker(self, worker_address, band_to_slot=None):
...
+
+        task = asyncio.create_task(_release_worker())
+        task.add_done_callback(lambda _: self._releasing_tasks.pop(address, None))
+        self._releasing_tasks[address] = task
+        return await task
+
+    async def reconstruct_worker(self, address: str):
+        task = self._releasing_tasks.get(address)
+        if task is not None:
+            raise Exception(f"Can't reconstruct releasing worker {address}")
```
[the remark this answers is not in the document]
Comment (wjsi) [reply mars/deploy/oscar/ray.py:207]: `ReconstructWorkerError`?

File: mars/oscar/backends/pool.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wjsi) [reply mars/oscar/backends/pool.py:1046]: Users need to be informed in time when unexpected errors occurs. More work is needed and a simple error log is far from complete.

File: mars/services/scheduling/worker/execution.py (hunk 1 above)
Comment (wjsi) [reply mars/services/scheduling/worker/execution.py:339]: Then I would prefer adding `session_stid` to `release_free_slot` method and assert equality inside it. I'd prefer raising errors at user end to fail silently.

File: mars/deploy/oscar/tests/test_fault_injection.py (hunk 3 above)
Comment (fyrestone) [reply mars/deploy/oscar/tests/test_fault_injection.py]: The print has been removed.

File: mars/services/scheduling/worker/workerslot.py (hunk 2 above)
Comment (fyrestone) [reply mars/services/scheduling/worker/workerslot.py]: I have added a `register_free_slot` API for the `BandSlotControlActor`.

File: mars/services/scheduling/worker/execution.py (hunk 1 above)
Comment (fyrestone) [reply mars/services/scheduling/worker/execution.py:339]: Got it. I added some checks to `release_free_slot`.

File: mars/deploy/oscar/ray.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fyrestone) [reply mars/deploy/oscar/ray.py]: New `ReconstructWorkerError` is added for the `reconstruct_worker`.

File: mars/services/scheduling/worker/workerslot.py [hunk 7]
```diff
@@ -134,31 +143,48 @@ async def acquire_free_slot(self, session_stid: Tuple[str, str]):
...
 
-        session_stid = self._slot_to_session_stid.pop(slot_id, None)
-        self._session_stid_to_slot.pop(session_stid, None)
         logger.debug('Slot %d released', slot_id)
 
         if slot_id not in self._free_slots:
             self._free_slots.add(slot_id)
             self._semaphore.release()
 
+    def register_free_slot(self, slot_id: int, pid: int):
```
Comment (wjsi) [inline mars/services/scheduling/worker/workerslot.py:163]: Rename to `register_slot` as registered slots may not always be free.

Comment (fyrestone) [reply mars/services/scheduling/worker/workerslot.py:163]: OK

**Review by wjsi — APPROVED** [review]
LGTM

**Review by qinxuye — APPROVED** [review]
LGTM
